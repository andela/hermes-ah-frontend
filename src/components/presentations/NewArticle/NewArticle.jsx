import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { EditorState, convertToRaw } from 'draft-js';
import { toast } from 'react-toastify';
import draftToHtml from 'draftjs-to-html';
import uploadToCloudnary from '../../../utils/uploadToCloudnary';
import NewArticleForm from './NewArticleForm/NewArticleForm';
import validateImage from '../../../utils/validateImage';
import NavBar from '../../shared/NavBar/NavBar';

class NewArticle extends Component {
  constructor() {
    super();
    this.state = {
      category: '',
      keywords: [],
      imageUrl: '',
      editorState: EditorState.createEmpty(),
      body: '',
      title: '',
      abstract: '',
    };
    this.saveOrPublish = this.saveOrPublish.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onEditorStateChange = this.onEditorStateChange.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value.toLowerCase() });
  }

  onEditorStateChange = editorState => {
    const body = draftToHtml(convertToRaw(editorState.getCurrentContent()));
    this.setState({
      body,
      editorState,
    });
  };

  saveToCloudinary = async e => {
    const form = new FormData();
    const imageData = e.target.files[0];
    const validFormat = validateImage(imageData);
    if (validFormat.valid) {
      toast.info(validFormat.message, {
        type: toast.TYPE.INFO,
        closeButton: false,
        position: toast.POSITION.TOP_CENTER,
      });
      form.append('file', imageData);
      const res = await uploadToCloudnary(form);
      this.setState({ imageUrl: res.url });
      toast.dismiss();
    } else {
      toast.error(validFormat.message);
      e.target.value = '';
    }
  };

  saveKeywords = e => {
    e.preventDefault();
    const keywordData = e.target.attributes.getNamedItem('name').value;
    const { keywords } = this.state;
    this.setState({ keywords: [...keywords, keywordData] });
  };

  saveCategory = e => {
    e.preventDefault();
    const categoryData = e.target.attributes.getNamedItem('name').value;
    this.setState({ category: categoryData });
  };

  saveOrPublish = async (e, isDraft) => {
    e.preventDefault();
    const articleInput = this.state;
    const { postArticle } = this.props;
    const { body, imageUrl } = this.state;
    const data = {
      title: articleInput.title,
      abstract: articleInput.abstract,
      is_draft: isDraft,
      image_url: imageUrl,
      body,
      keywords: articleInput.keywords,
      category: articleInput.category,
    };
    await postArticle(data);
  };

  render() {
    const { editorState } = this.state;
    return (
      <React.Fragment>
        <NavBar />
        <NewArticleForm
          saveCategory={this.saveCategory}
          saveKeywords={this.saveKeywords}
          onEditorStateChange={this.onEditorStateChange}
          editorState={editorState}
          onChange={this.onChange}
          saveOrPublish={this.saveOrPublish}
          saveToCloudinary={this.saveToCloudinary}
        />
      </React.Fragment>
    );
  }
}

NewArticle.propTypes = {
  postArticle: PropTypes.func.isRequired,
};

export default NewArticle;
