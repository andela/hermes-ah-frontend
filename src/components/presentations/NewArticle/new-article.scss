@import '../../../assets/stylesheets/index.scss';

p{
    text-align: left;
}
.title{
    background-color: black;
    height: 300px;
}

.editor{
    border: 3px solid #0000;
}

.DraftEditor-root {
    background-color: #ffffff;
    border: 1px solid rgb(192, 192, 192);
    width:100%;
    height:400px;
    overflow-y: scroll;
    font-size: 18px;
    font-family: 'calibri', sans-serif;
    border-radius: 5px;
  }

//   .DraftEditor-editorContainer {
//     //   border: black solid 1px;
//     //   height: 80px;
//   }

  .keywordTitle {
      color: #857E7E;
  }

  .labels {
    font-size: 2vh;
    font-weight: bold;
  }

  .inputBox {
      width: 90%;
  }
  .inputBox:nth-child(2) {
    width: 60%;
  }
  .whole-text {
      margin: 0 10% 0 8%;
  }

  .image-upload {
    padding: 1% 0 0 2%;
    width: 35%;
    height: 130px;
    border: 1px solid #e7e6e6;
  }

  .image-upload p {
    margin: 0;
  }

  .image-upload-box {
    background-color: #EBEBEB;
    padding: 0 auto;
    width: 80%;
    height: 50px;
    border: 1px solid #c0bfbf;
}

.image-upload-box input[type="file"] {
    margin: 14px 16% 0;
}

.ui.button{
   background-color: $primary-color;
   color: #ffffff;
}

.ui.basic.blue.button {
    color: #2F80ED!important; 
}
button{
    padding: 1%;
    border: 1px solid rgb(221, 221, 221);
}

button:hover{
    background-color: #e7e6e6;
}
