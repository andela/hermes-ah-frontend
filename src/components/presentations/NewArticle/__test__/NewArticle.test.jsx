import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import NewArticle from '../NewArticle';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const store = mockStore({});

const NewArticleComponent = (
  <Provider store={store}>
    <NewArticle createNewArticle="article" />
  </Provider>
);

describe('NewArticle component', () => {
  it('should match snapshot', () => {
    const wrap = NewArticleComponent;
    expect(wrap).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    const wrapper = shallow(NewArticleComponent);
    expect(wrapper.find('div'));
  });
});
