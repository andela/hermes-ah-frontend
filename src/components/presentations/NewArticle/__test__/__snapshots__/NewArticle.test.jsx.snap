// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewArticle component should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NewArticle
    createNewArticle="article"
  />
</Provider>
`;
